rfd/0010-api.md

```diff
@@ -35,89 +34,155 @@ func connectClient() (*auth.Client, error) {
+
+For example, in `lib/services/role`, `role.CheckAndSetDefaults` has some business logic that should remain on the server side. Therefore that logic should be extracted into a new function, `ValidateRole`, and kept in `lib/services/role.go` rather than `api/types/role.go`.
+
+```go
+// api/types/role.go
+(r *RoleV3) CheckAndSetDefaults() error {
+  ... basic validation with business logic extracted
+}
 ```
 
-#### Move client implementation to `api/client`
+```go
+// lib/services/role.go
+func ValidateRole(r types.Role) error {
+  ... business logic
+}
+```
+
+Some basic dependencies will be unavoidable:
+  * `gravitational/trace`
+  * `google.golang.org/grpc`
+  * `golang.org/x/net`
+  * `golang.org/x/crypto`
+  * `github.com/stretchr/testify`
+
+If the list of dependencies grows beyond 20 dependencies and indirect dependencies, this issue should be revisited for better tracking mechanisms.
 
-Move relevant lib/client.*Client code to `api/client`.
-Track dependencies, make sure that new `api/client` module has no dependencies
-on the outside modules anywhere in `lib`.
+##### Deprecate the `lib/auth.Client`
 
-**Track dependencies**
+As stated before, all gRPC client related code will be moved out of `teleport/lib/auth/clt.go` and into `api/client`. Backwards compatibility will be maintained by embedding the gRPC `api/client.Client` in the internal `teleport/lib/auth.Client`.
 
-Track dependencies to be limited to:
-  * `gravitational/roundtrip`,  `gravitational/trace`
-  * grpc google package and its dependencies
-  * protobuf generated code
-  * Keep some auth service specific code like `utils.Addr` out of the `api/client`
+Additionally, the remaining HTTP endpoints in the internal client will be converted and added to the gRPC client. To maintain backwards compatibility with old servers, the internal client will be updated to first try the new gRPC endpoint before falling back to the existing HTTP endpoint. These fallback methods will be moved to `teleport/lib/auth/httpfallback.go` and deprecated after one major release cycle.
 
-**Go native user experience**
+There is [issue](https://github.com/gravitational/teleport/issues/6394) used to track remaining endpoints that need to be converted to gRPC. 
 
-Achieve the user experience in Client with a simplified client constructor.
-This should also transparently handle dialing via a proxy.
+Once all necessary endpoints have been added in gRPC and all HTTP fallback logic has been deprecated and removed, the gRPC client can completely replace the existing `lib/auth.Client`. At this point the underlying gRPC server will be the same, so the transition should be relatviley straightforward.
+
+### api module
+
+The api package will be made a sub module of teleport. It will exist in the `teleport/api`, and will be imported by the main `teleport` module through a go.mod [replace directive](https://golang.org/ref/mod#go-mod-file-replace) using a relative path.
```

Comment:
```suggestion
The api package will be made a sub module of teleport. It will exist in the `teleport/api` directory, and will be imported by the main `teleport` module through a go.mod [replace directive](https://golang.org/ref/mod#go-mod-file-replace) using a relative path.
```